### PR TITLE
[CLN] More cython cleanups, with bonus type annotations

### PR DESCRIPTION
**Pull request by jbrockmendel**
A few places where string formatting is modernized

Removed remaining `# cython: profile=False` comments

The main (and possible controversial) thing done here is changing some functions to use type-hint syntax.  The affected functions are all currently `cpdef`, but are never used in cython, so _should_ only be `def`.  But cython's syntax does not allow for specifying a return type in `def` functions, so this is the cleanest way to remove unnecessary `cpdef` without losing typing.  (some discussion of this occurred in #22180)

### CODE REVIEW

**Comment by WillAyd** [comment]
Out of curiosity how are you verifying that the `cpdef` functions are only being called from Python? Manual exercise or is there a tool to help identify?

**Comment by jbrockmendel** [comment]
> Out of curiosity how are you verifying that the cpdef functions are only being called from Python? Manual exercise or is there a tool to help identify?

Just text search within the file.  Since the relevant functions aren't exposed in a .pxd file, there is no risk of them being called from a different cython file.

**Comment by jbrockmendel** [comment]
@jreback would this be easier if I separated the no-profile comments into an isolated PR?

File: pandas/_libs/writers.pyx [hunk 1]
```diff
@@ -124,19 +123,19 @@ def convert_json_to_lines(object arr):
...
+def max_len_string_array(pandas_string[:] arr) -> Py_ssize_t:
     """ return the maximum size of elements in a 1-dim string array """
     cdef:
         Py_ssize_t i, m = 0, l = 0, length = arr.shape[0]
         pandas_string v
 
     for i in range(length):
         v = arr[i]
-        if PyString_Check(v):
+        if isinstance(v, str):
```
Comment (jreback) [inline pandas/_libs/writers.pyx]: do these make any perf diffs?

**Review by jreback — CHANGES_REQUESTED** [review]
looks ok, can you rebase

File: pandas/_libs/writers.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/writers.pyx]: cython makes this substitution on its own

**Comment by jbrockmendel** [comment]
@jreback gentle ping; got your OK last week

**Comment by jreback** [comment]
think this was ok, can you rebase

**Comment by jbrockmendel** [comment]
Rebased; circleCI fail is hypothesis timeout

**Comment by jbrockmendel** [comment]
Rebased+green

File: pandas/_libs/algos_common_helper.pxi.in [hunk 2]
```diff
@@ -587,7 +587,7 @@ def get_dispatch(dtypes):
 
 {{for name, c_type, dtype in get_dispatch(dtypes)}}
 
-cpdef ensure_{{name}}(object arr, copy=True):
+def ensure_{{name}}(object arr, copy=True):
```
Comment (jreback) [inline pandas/_libs/algos_common_helper.pxi.in]: are these for sure not called in cython code?

File: pandas/_libs/hashing.pyx [hunk 3]
```diff
@@ -132,6 +132,7 @@ cdef inline void _sipround(uint64_t* v0, uint64_t* v1,
     v2[0] = _rotl(v2[0], 32)
 
 
+# TODO: This appears unused; remove?
```
Comment (jreback) [inline pandas/_libs/hashing.pyx]: might be unused - it was a part of the hashing at one point iirc

File: pandas/_libs/lib.pyx [hunk 4]
```diff
@@ -107,7 +107,7 @@ def memory_usage_of_objects(object[:] arr):
 # ----------------------------------------------------------------------
 
 
-cpdef bint is_scalar(object val):
+def is_scalar(val: object) -> bint:
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: this not ever called in cython?

File: pandas/_libs/algos_common_helper.pxi.in (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_common_helper.pxi.in]: non-cimported only

File: pandas/_libs/lib.pyx (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx]: Yep, easy to confirm via grep

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**